pennylane/data/dataset.py

```diff
@@ -23,25 +23,34 @@
 
 class Dataset(ABC):
     "The dataset data type. Allows users to create datasets with their own data"
+    CFG_ATTRS = {"dtype", "folder", "file_prefix", "file_fstring", "_attributes", "__doc__"}
 
-    def __init__(self, dfile=None, dtype=None, **kwargs):
-        self.dfile = dfile
+    def __init__(self, dtype, folder, file_prefix, attributes=None, **kwargs):
         self.dtype = dtype
+        self.folder = folder
+        self.file_prefix = file_prefix
+        self.file_fstring = os.path.join(folder, file_prefix) + "_{}.dat"
+        self._attributes = attributes or []
         self.__doc__ = ""
         for key, val in kwargs.items():
             setattr(self, f"{key}", val)
 
+    def __get_filename_for_attribute(self, attribute):
+        return self.file_fstring.format(attribute)
+
     def __getattribute__(self, name):
         try:
-            value = super().__getattribute__(name)
+            return super().__getattribute__(name)
         except AttributeError:
-            value = None
-        dfile = super().__getattribute__("dfile")
-        if value is None and dfile is not None:
-            file_path = f"{dfile}_{name}.dat"
-            if os.path.exists(file_path):
-                value = self._read_file(file_path)
-        return value
+            pass
+        filepath = self.__get_filename_for_attribute(name)
```

Comment:
I changed the code to check if `_full.dat` exists first. Note that if an attribute isn't there, but it _does_ have its own file, it will be missed. We can check both, but I think something weirder is going on if it's not in the full file.